File: 251__Flatten_2D_Vector.py
class Vector2D(object):

    def __init__(self, vec2d):
        """
        Initialize your data structure here.
        :type vec2d: List[List[int]]
        """
        self.list = []
        for lst in vec2d:
            self.list += lst
        self.length = len(self.list)
        self.index = -1

    def next(self):
        """
        :rtype: int
        """
        self.index += 1
        return self.list[self.index]

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.index + 1 < self.length

# Your Vector2D object will be instantiated and called as such:
# i, v = Vector2D(vec2d), []
# while i.hasNext(): v.append(i.next())

File: Java/251__Flatten_2D_Vector.java
class Vector2D {

    private int[][] v = null;
    private int i = 0, j = 0;

    public Vector2D(int[][] v) {
        this.v = v;
    }

    public int next() { // hasNext() finds the next valid position
        return hasNext() ? v[i][j++] : -1;
    }

    public boolean hasNext() {
        while (i < v.length && j >= v[i].length) {
            i++;
            j = 0;
        }
        return i != v.length;
    }
}

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D obj = new Vector2D(v);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */
 
